File: packages/python/src/quanttide_agent/llm.py
from __future__ import annotations

from typing import Any, Literal

import anyio
import httpx

from .config import settings
from .cost import Usage
from .message import ChatResponse, Message
from .tool import ToolCall, ToolSchema


class LLMError(Exception):
    """Raised when LLM chat fails after retries.

    >>> issubclass(LLMError, Exception)
    True
    """


class BaseLLM:
    """Shared initialization and chat helpers for sync/async LLM clients."""

    def __init__(
        self,
        model: str | None = None,
        base_url: str | None = None,
        api_key: str | None = None,
        *,
        _client: httpx.Client | httpx.AsyncClient | None = None,
    ):
        self.model = model or settings.llm_model
        api_key = api_key or settings.llm_api_key
        base_url = (base_url or settings.llm_base_url).rstrip("/")
        self._client: httpx.Client | httpx.AsyncClient = _client

    @staticmethod
    def build_chat_body(
        messages: list[Message] | list[dict] | str,
        *,
        model: str,
        temperature: float | None = None,
        max_tokens: int | None = None,
        top_p: float | None = None,
        stop: str | list[str] | None = None,
        frequency_penalty: float | None = None,
        presence_penalty: float | None = None,
        thinking: bool | None = None,
        reasoning_effort: Literal["low", "medium", "high", "max"] | None = None,
        tools: list[ToolSchema] | None = None,
        tool_choice: str | None = None,
        response_format: dict | None = None,
    ) -> dict:
        if isinstance(messages, str):
            body_messages: list[dict] = [{"role": "user", "content": messages}]
        elif messages and isinstance(messages[0], Message):
            body_messages = [m.to_dict() for m in messages]
        else:
            body_messages = messages  # type: ignore

        body: dict[str, Any] = {"model": model, "messages": body_messages}

        _params: dict[str, Any] = {
            "temperature": temperature,
            "max_tokens": max_tokens,
            "top_p": top_p,
            "stop": stop,
            "frequency_penalty": frequency_penalty,
            "presence_penalty": presence_penalty,
            "reasoning_effort": reasoning_effort,
            "tool_choice": tool_choice,
            "response_format": response_format,
        }
        body.update({k: v for k, v in _params.items() if v is not None})

        if thinking is not None:
            body["thinking"] = {"type": "enabled" if thinking else "disabled"}
        if tools is not None:
            body["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": t.name,
                        "description": t.description,
                        "parameters": t.parameters
                        or {"type": "object", "properties": {}},
                    },
                }
                for t in tools
            ]

        return body

    @staticmethod
    def parse_chat_response(data: dict, model: str) -> ChatResponse:
        choice = data["choices"][0]
        msg = choice["message"]

        tool_calls = None
        if msg.get("tool_calls"):
            tool_calls = [
                ToolCall(
                    id=tc["id"],
                    name=tc["function"]["name"],
                    arguments=tc["function"]["arguments"],
                )
                for tc in msg["tool_calls"]
            ]

        usage_raw = data.get("usage")
        usage = Usage.from_api(usage_raw) if usage_raw else None

        return ChatResponse(
            content=msg.get("content", "") or "",
            model=data.get("model", model),
            finish_reason=choice.get("finish_reason", "stop"),
            reasoning_content=msg.get("reasoning_content"),
            tool_calls=tool_calls,
            usage=usage,
        )


class LLM(BaseLLM):
    """Sync LLM client.

    Usage::

        llm = LLM(model="deepseek-v4-pro", api_key="sk-...")
        resp = llm.complete("Hello")
        print(resp.content)
    """

    def __init__(
        self,
        model: str | None = None,
        base_url: str | None = None,
        api_key: str | None = None,
        *,
        _http_client: httpx.Client | None = None,
    ):
        api_key = api_key or settings.llm_api_key
        base_url = (base_url or settings.llm_base_url).rstrip("/")
        super().__init__(
            model=model,
            _client=_http_client
            or httpx.Client(
                base_url=base_url,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                timeout=120,
            ),
        )

    def complete(
        self,
        messages: list[Message] | list[dict] | str,
        *,
        model: str | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
        top_p: float | None = None,
        stop: str | list[str] | None = None,
        frequency_penalty: float | None = None,
        presence_penalty: float | None = None,
        thinking: bool | None = None,
        reasoning_effort: Literal["low", "medium", "high", "max"] | None = None,
        tools: list[ToolSchema] | None = None,
        tool_choice: str | None = None,
        response_format: dict | None = None,
        retry: int = 0,
    ) -> ChatResponse:
        body = self.build_chat_body(
            messages,
            model=model or self.model,
            temperature=temperature,
            max_tokens=max_tokens,
            top_p=top_p,
            stop=stop,
            frequency_penalty=frequency_penalty,
            presence_penalty=presence_penalty,
            thinking=thinking,
            reasoning_effort=reasoning_effort,
            tools=tools,
            tool_choice=tool_choice,
            response_format=response_format,
        )

        last_error: Exception | None = None
        for _ in range(max(retry + 1, 1)):
            try:
                resp = self._client.post("/chat/completions", json=body)  # type: ignore[union-attr]
                resp.raise_for_status()
                data: dict = resp.json()
                break
            except httpx.HTTPStatusError as e:
                last_error = e
                continue
        else:
            assert last_error is not None
            raise LLMError("chat failed after retries") from last_error

        return self.parse_chat_response(data, model or self.model)


class AsyncLLM(BaseLLM):
    """Async LLM client.

    Usage::

        llm = AsyncLLM(model="deepseek-v4-pro", api_key="sk-...")
        resp = await llm.complete("Hello")
        print(resp.content)
    """

    def __init__(
        self,
        model: str | None = None,
        base_url: str | None = None,
        api_key: str | None = None,
        *,
        _http_client: httpx.AsyncClient | None = None,
    ):
        api_key = api_key or settings.llm_api_key
        base_url = (base_url or settings.llm_base_url).rstrip("/")
        super().__init__(
            model=model,
            _client=_http_client
            or httpx.AsyncClient(
                base_url=base_url,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                timeout=120,
            ),
        )

    async def complete(
        self,
        messages: list[Message] | list[dict] | str,
        *,
        model: str | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
        top_p: float | None = None,
        stop: str | list[str] | None = None,
        frequency_penalty: float | None = None,
        presence_penalty: float | None = None,
        thinking: bool | None = None,
        reasoning_effort: Literal["low", "medium", "high", "max"] | None = None,
        tools: list[ToolSchema] | None = None,
        tool_choice: str | None = None,
        response_format: dict | None = None,
        retry: int = 0,
    ) -> ChatResponse:
        body = self.build_chat_body(
            messages,
            model=model or self.model,
            temperature=temperature,
            max_tokens=max_tokens,
            top_p=top_p,
            stop=stop,
            frequency_penalty=frequency_penalty,
            presence_penalty=presence_penalty,
            thinking=thinking,
            reasoning_effort=reasoning_effort,
            tools=tools,
            tool_choice=tool_choice,
            response_format=response_format,
        )

        last_error: Exception | None = None
        for _ in range(max(retry + 1, 1)):
            try:
                resp = await self._client.post("/chat/completions", json=body)
                resp.raise_for_status()
                data: dict = resp.json()
                break
            except httpx.HTTPStatusError as e:
                last_error = e
                continue
        else:
            assert last_error is not None
            raise LLMError("chat failed after retries") from last_error

        return self.parse_chat_response(data, model or self.model)

File: packages/rust/src/llm.rs
use std::fmt;

use serde_json::Value;

use crate::config::Settings;
use crate::cost::Usage;
use crate::message::{ChatResponse, Message};
use crate::tool::{ToolCall, ToolSchema};

#[derive(Debug)]
pub struct LLMError(pub String);

impl fmt::Display for LLMError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "LLM error: {}", self.0)
    }
}

impl std::error::Error for LLMError {}

pub trait HttpClient: Send + Sync {
    fn post_json(&self, url: &str, auth: &str, body: &Value) -> Result<Value, LLMError>;
}

pub struct UreqClient;

impl HttpClient for UreqClient {
    fn post_json(&self, url: &str, auth: &str, body: &Value) -> Result<Value, LLMError> {
        let resp = ureq::post(url)
            .set("Authorization", auth)
            .set("Content-Type", "application/json")
            .send_json(body)
            .map_err(|e| LLMError(format!("request failed: {}", e)))?;
        resp.into_json()
            .map_err(|e| LLMError(format!("parse failed: {}", e)))
    }
}

pub struct LLM {
    model: String,
    base_url: String,
    api_key: String,
    client: Box<dyn HttpClient>,
}

impl Default for LLM {
    fn default() -> Self {
        let settings = Settings::from_env();
        Self::new(&settings.llm_model, &settings.llm_base_url, &settings.llm_api_key)
    }
}

impl LLM {
    pub fn new(model: &str, base_url: &str, api_key: &str) -> Self {
        Self {
            model: model.to_string(),
            base_url: base_url.trim_end_matches('/').to_string(),
            api_key: api_key.to_string(),
            client: Box::new(UreqClient),
        }
    }

    pub fn with_client(model: &str, base_url: &str, api_key: &str, client: Box<dyn HttpClient>) -> Self {
        Self {
            model: model.to_string(),
            base_url: base_url.trim_end_matches('/').to_string(),
            api_key: api_key.to_string(),
            client,
        }
    }

    pub fn complete(
        &self,
        messages: &[Message],
        options: CompleteOptions,
    ) -> Result<ChatResponse, LLMError> {
        let body = self.build_body(messages, &options);
        let url = format!("{}/chat/completions", self.base_url);
        let auth = format!("Bearer {}", self.api_key);
        let data = self.client.post_json(&url, &auth, &body)?;
        parse_response(&data, &self.model)
    }

    fn build_body(&self, messages: &[Message], options: &CompleteOptions) -> Value {
        let msg_list: Vec<Value> = messages.iter().map(|m| m.to_dict()).collect();

        let mut body = serde_json::Map::new();
        body.insert("model".to_string(), Value::String(options.model.clone().unwrap_or_else(|| self.model.clone())));
        body.insert("messages".to_string(), Value::Array(msg_list));

        if let Some(t) = options.temperature {
            body.insert("temperature".to_string(), Value::Number(serde_json::Number::from_f64(t).unwrap()));
        }
        if let Some(m) = options.max_tokens {
            body.insert("max_tokens".to_string(), Value::Number(m.into()));
        }
        if let Some(p) = options.top_p {
            body.insert("top_p".to_string(), Value::Number(serde_json::Number::from_f64(p).unwrap()));
        }
        if let Some(s) = &options.stop {
            match s {
                Stop::Single(st) => body.insert("stop".to_string(), Value::String(st.clone())),
                Stop::List(list) => body.insert("stop".to_string(), Value::Array(list.iter().map(|s| Value::String(s.clone())).collect())),
            };
        }
        if let Some(fp) = options.frequency_penalty {
            body.insert("frequency_penalty".to_string(), Value::Number(serde_json::Number::from_f64(fp).unwrap()));
        }
        if let Some(pp) = options.presence_penalty {
            body.insert("presence_penalty".to_string(), Value::Number(serde_json::Number::from_f64(pp).unwrap()));
        }
        if let Some(re) = &options.reasoning_effort {
            body.insert("reasoning_effort".to_string(), Value::String(re.clone()));
        }
        if let Some(tc) = &options.tool_choice {
            body.insert("tool_choice".to_string(), Value::String(tc.clone()));
        }
        if let Some(rf) = &options.response_format {
            body.insert("response_format".to_string(), rf.clone());
        }
        if let Some(thinking) = options.thinking {
            body.insert("thinking".to_string(), serde_json::json!({"type": if thinking { "enabled" } else { "disabled" }}));
        }
        if let Some(tools) = &options.tools {
            let tool_list: Vec<Value> = tools
                .iter()
                .map(|t| {
                    serde_json::json!({
                        "type": "function",
                        "function": {
                            "name": t.name,
                            "description": t.description,
                            "parameters": t.parameters.as_ref().unwrap_or(&serde_json::json!({"type": "object", "properties": {}})),
                        }
                    })
                })
                .collect();
            body.insert("tools".to_string(), Value::Array(tool_list));
        }

        Value::Object(body)
    }
}

pub fn parse_response(data: &Value, default_model: &str) -> Result<ChatResponse, LLMError> {
    let choice = data["choices"][0].clone();
    let msg = choice["message"].clone();

    let tool_calls = msg["tool_calls"].as_array().map(|arr| {
        arr.iter()
            .map(|tc| ToolCall {
                id: tc["id"].as_str().unwrap_or("").to_string(),
                name: tc["function"]["name"].as_str().unwrap_or("").to_string(),
                arguments: tc["function"]["arguments"].as_str().unwrap_or("").to_string(),
            })
            .collect()
    });

    let usage_raw = data.get("usage");
    let usage = usage_raw.and_then(Usage::from_api);

    Ok(ChatResponse {
        content: msg["content"].as_str().unwrap_or("").to_string(),
        model: data["model"].as_str().unwrap_or(default_model).to_string(),
        finish_reason: choice["finish_reason"].as_str().unwrap_or("stop").to_string(),
        reasoning_content: msg["reasoning_content"].as_str().map(String::from),
        tool_calls,
        usage,
    })
}

pub struct CompleteOptions {
    pub model: Option<String>,
    pub temperature: Option<f64>,
    pub max_tokens: Option<u32>,
    pub top_p: Option<f64>,
    pub stop: Option<Stop>,
    pub frequency_penalty: Option<f64>,
    pub presence_penalty: Option<f64>,
    pub thinking: Option<bool>,
    pub reasoning_effort: Option<String>,
    pub tools: Option<Vec<ToolSchema>>,
    pub tool_choice: Option<String>,
    pub response_format: Option<Value>,
}

impl Default for CompleteOptions {
    fn default() -> Self {
        Self {
            model: None,
            temperature: None,
            max_tokens: None,
            top_p: None,
            stop: None,
            frequency_penalty: None,
            presence_penalty: None,
            thinking: None,
            reasoning_effort: None,
            tools: None,
            tool_choice: None,
            response_format: None,
        }
    }
}

/// Parse structured output from LLM response text.
///
/// Handles markdown code blocks (` ```json `, ` ```yaml `),
/// raw JSON objects `{}`, and raw JSON arrays `[]`.
pub fn parse_structured_output(response: &str) -> Result<Value, String> {
    for marker in &["```json", "```JSON", "```yaml", "```YAML"] {
        if let Some(start) = response.find(marker) {
            let s = start + marker.len();
            let e = response[s..].find("```").map(|i| s + i).unwrap_or(response.len());
            let trimmed = response[s..e].trim();
            if let Ok(v) = serde_json::from_str(trimmed) {
                return Ok(v);
            }
        }
    }
    if let Some(start) = response.find('{') {
        let e = response.rfind('}').map(|i| i + 1).unwrap_or(response.len());
        let trimmed = response[start..e].trim();
        if let Ok(v) = serde_json::from_str(trimmed) {
            return Ok(v);
        }
    }
    if let Some(start) = response.find('[') {
        let e = response.rfind(']').map(|i| i + 1).unwrap_or(response.len());
        let trimmed = response[start..e].trim();
        if let Ok(v) = serde_json::from_str(trimmed) {
            return Ok(v);
        }
    }
    Err("No valid JSON found in response".to_string())
}

pub enum Stop {
    Single(String),
    List(Vec<String>),
}
